aesd-char-driver/main.c

```diff
@@ -243,14 +246,136 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+    //      and we don't have a "pop" or "remove" function for our circular buffer. So entries always either
+    //      have a valid size or a 0 size. 
+    AESD_CIRCULAR_BUFFER_FOREACH(entry,&(dev->circ_buff),index) {
+        if (entry != NULL)
+            total_buff_bytes += entry->size;
+    }
+
+    PDEBUG("Total bytes in buffer to seek: %lld", total_buff_bytes);
+    
+    //return if mutex wait interrupted
+    if (mutex_lock_interruptible(&dev->lock))
+        return -ERESTARTSYS;
+
+    retval = fixed_size_llseek(filp, offset, whence, total_buff_bytes);
+
+    mutex_unlock(&dev->lock);
+    return retval;
+}
+
+//New for assignment 9: ioctl implementation and helper function
+//Reference: scull character driver main.c scull_ioctl
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset){
+    struct aesd_dev *dev = filp->private_data;
+    unsigned int num_cmds = 0; //Total number of strings in the circ buffer
+    unsigned int buff_idx = 0; //Location in the circ buffer of target string
+    unsigned int curr_idx = 0; //index used to iterate through circular buffer
+    unsigned int i = 0; 
+    ssize_t total_pos = 0; //the new offset for f_pos
+    struct aesd_buffer_entry target_entry; //Entry in circ buffer containing target string
+    struct aesd_buffer_entry curr_entry; //current entry when iterateing circular buffer
+
+    if (dev->circ_buff.full)
+        num_cmds = AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+    else
+        //Reference: Howdy Pierce 2022 PES data structures lecture
+        num_cmds = ( (AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED + dev->circ_buff.in_offs - dev->circ_buff.out_offs) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED);
+
+    if (write_cmd > (num_cmds -1) ) //-1 because write_cmd is 0 indexed 
+        return -EINVAL;
+    if (write_cmd > AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
+        return -EINVAL;
+
+    PDEBUG("%u commands found in buffer. %u is a legal command index", num_cmds, write_cmd);
+
+    //Find the index in the circular buffer corresponding to the requested cmd
+    buff_idx = ( (dev->circ_buff.out_offs + write_cmd) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED ); 
+    target_entry = dev->circ_buff.entry[buff_idx];
+
+    if (write_cmd_offset >= target_entry.size) //>= becuase write_cmd_offset is 0 indexed
+        return -EINVAL;
+
+    //Iterate through all valid commands (besides target one) to get total offset
+    curr_idx = dev->circ_buff.out_offs;
+    for (i = 0; i < write_cmd; i++){ 
+        curr_entry = dev->circ_buff.entry[curr_idx];
+        total_pos += curr_entry.size; 
+        curr_idx = ( (curr_idx + 1) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED);
+    }
+
+    total_pos += write_cmd_offset; //finally add in the offset within the target command
+
+    //return if mutex wait interrupted
+    if (mutex_lock_interruptible(&dev->lock))
+        return -ERESTARTSYS;
+
+    filp->f_pos = total_pos;
```

Comment:
@malcolmmckellips I don't think this locking interval is what you want either.  What is `dev->lock` locking?  Is `filp->f_pos` protected by `dev->lock`?  See discussion in class about this.